// @hashlips-lab/art-engine/dist/plugins/renderers/ItemAttributesRenderer.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ItemAttributesRenderer = void 0;
const ItemAttributesRendererInterface_1 = require("../../interfaces/renderers/ItemAttributesRendererInterface");
const ItemAttributesGeneratorInterface_1 = require("../../interfaces/generators/ItemAttributesGeneratorInterface");
class ItemAttributesRenderer {
    attributesGetter;
    name;
    description;
    constructor(constructorProps = {}) {
        this.name = constructorProps.name ? constructorProps.name : () => "";
        this.description = constructorProps.description
            ? constructorProps.description
            : () => "";
    }
    async init(props) {
        this.attributesGetter = props.attributesGetter;
    }
    async render() {
        const renders = {};
        for (const [itemUid, attributes] of Object.entries(this.attributesGetter())) {
            const supportedAttributes = attributes
                .filter((attribute) => ItemAttributesGeneratorInterface_1.ITEM_ATTRIBUTES_GENERATOR_INTERFACE_V1 === attribute.kind)
                .map((attribute) => attribute.data);
            if (supportedAttributes.length < 1) {
                throw new Error(`Couldn't find any supported set of attributes for the current item: ${itemUid}`);
            }
            renders[itemUid] = [
                supportedAttributes.reduce((mergedAttributes, newAttributes) => {
                    mergedAttributes.data.dna.push(newAttributes.dna);
                    mergedAttributes.data.name = this.name(itemUid);
                    for (const key in newAttributes.attributes) {
                        if (mergedAttributes.data.attributes[key] === undefined) {
                            mergedAttributes.data.attributes[key] = [];
                        }
                        mergedAttributes.data.attributes[key].push(newAttributes.attributes[key]);
                    }
                    mergedAttributes.data.description = this.description(mergedAttributes.data.attributes);
                    return mergedAttributes;
                }, {
                    kind: ItemAttributesRendererInterface_1.ITEM_ATTRIBUTES_RENDERER_INTERFACE_V1,
                    data: {
                        dna: [],
                        name: "",
                        description: "",
                        attributes: {},
                    },
                }),
            ];
        }
        return renders;
    }
}
exports.ItemAttributesRenderer = ItemAttributesRenderer;
