// @hashlips-lab/art-engine/dist/core/ItemsDataManager.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
class ItemsDataManager {
    areAttributesReadOnlyMode;
    areRendersReadOnlyMode;
    itemsAttributes;
    itemsRenders;
    constructor() {
        this.areAttributesReadOnlyMode = false;
        this.areRendersReadOnlyMode = false;
        this.itemsAttributes = {};
        this.itemsRenders = {};
    }
    freezeAttributes() {
        this.areAttributesReadOnlyMode = true;
    }
    freezeRenders() {
        this.areRendersReadOnlyMode = true;
    }
    getAttributes() {
        return this.itemsAttributes;
    }
    addAttributes(itemUid, attributes) {
        if (this.areAttributesReadOnlyMode) {
            throw new Error(`Trying to add attributes data on a frozen ItemsDataManager (item: "${itemUid}")`);
        }
        this.itemsAttributes[itemUid] = (this.itemsAttributes[itemUid] ?? []).concat(attributes);
    }
    getRenders() {
        return this.itemsRenders;
    }
    addRenders(itemUid, renders) {
        if (this.areRendersReadOnlyMode) {
            throw new Error(`Trying to add renders data on a frozen ItemsDataManager (item: "${itemUid}")`);
        }
        this.itemsRenders[itemUid] = (this.itemsRenders[itemUid] ?? []).concat(renders);
    }
}
exports.default = ItemsDataManager;
